pkg/kubelet/kubelet.go

```diff
@@ -2276,9 +2277,10 @@ func (kl *Kubelet) deletePod(pod *v1.Pod) error {
 func (kl *Kubelet) rejectPod(pod *v1.Pod, reason, message string) {
 	kl.recorder.Eventf(pod, v1.EventTypeWarning, reason, message)
 	kl.statusManager.SetPodStatus(pod, v1.PodStatus{
-		Phase:   v1.PodFailed,
-		Reason:  reason,
-		Message: "Pod was rejected: " + message})
+		QOSClass: v1qos.GetPodQOS(pod),
```

Comment:
since this test is not specific to QoS, I suggest this needs to be tested rather generically using e2e tests. I suggest the following test (under `test/e2e/node`, not the `test/e2e_node`):

1. Create a pod that requests too much memory so it is not scheduleable
2. Read the Pod status. It will be pending, but will have QoS set.
3. Set the `nodeName` for the Pod to that node explicitly to override the scheduler's checks
4. Check the status of a Pod that failed admission. Compare it with the pod status from before it was scheduled. Here I would think we can have an allowlist of fields to NOT compare - failure message and phase. But compare all other fields. This way we can catch "future" fields mutations.

The huge pod is a simplest way I know to make pod not admissible. You may think of other ways. But I think this is good enough.

Reply:
@SergeyKanzhelev  I added the test to the file `test/e2e/common/node/pod_admission.go`. The test is copied from the existing test https://github.com/carlory/kubernetes/blob/fix-126662-kubelet/test/e2e/common/node/pod_admission.go#L36, but I added a check to verify that the pod status's others field has not changed and I remove the `framework.WithNodeConformance()`.